src/tribol/physics/Mortar.cpp

```diff
@@ -631,6 +641,565 @@ void ComputeSingleMortarJacobian( SurfaceContactElem& elem )
+
+    // the sub-triangle is inverted.  likely something went wrong with CG.  don't try to integrate over it.
+    if ( area <= 0.0 ) {
+      continue;
+    }
+
+    for ( int j{ 0 }; j < 6; ++j ) {
+      RealT tri_xi[2] = { base_rule_2d[j * 2 + 0], base_rule_2d[j * 2 + 1] };
+      RealT tri_phi[3] = { 0.0, 0.0, 0.0 };
+      LinIsoTriShapeFunc( tri_xi, tri_phi );
+      RealT tri_quad_pt[2] = { tri_phi[0] * tri_0[0] + tri_phi[1] * tri_1[0] + tri_phi[2] * tri_2[0],
+                               tri_phi[0] * tri_0[1] + tri_phi[1] * tri_1[1] + tri_phi[2] * tri_2[1] };
+
+      // 3. map sub-triangle coordinate to nonmortar and mortar coordinates
+      // NOTE: we ideally want to do this in 2d, but there are finite differencing errors when we do.  The commented out
+      // version below does it in 2d.
+      RealT tri_quad_pt_3d[3] = { 0.0, 0.0, 0.0 };
+      Coords2DToPlane( tri_quad_pt, tri_quad_pt + 1, x0, e1, e2, tri_quad_pt_3d, 1 );
+      RealT xi1[2] = { 0.0, 0.0 };
+      InvIso( tri_quad_pt_3d, x1t, x1t + size1, x1t + 2 * size1, size1, xi1 );
+      RealT xi2[2] = { 0.0, 0.0 };
+      InvIso( tri_quad_pt_3d, x2t, x2t + size2, x2t + 2 * size2, size2, xi2 );
+      // 2D version begin...
+      // RealT xi1[2] = { 0.0, 0.0 };
+      // // NOTE: this limits routine to quads
+      // QuadInvIso( tri_quad_pt, x1t_2d, y1t_2d, xi1 );
+      // RealT xi2[2] = { 0.0, 0.0 };
+      // // NOTE: this limits routine to quads
+      // QuadInvIso( tri_quad_pt, x2t_2d, y2t_2d, xi2 );
+      // 2D version end...
+
+      RealT quad_wt = base_weights[j] * area;
+
+      // 4. Evaluate mortar matrix (nonmortar/nonmortar contribs)
+      // NOTE: local node numbering appears to be different using InvIso vs. QuadInvIso.  Since we are using InvIso, use
+      // the node numbering that appears to be consistent with that routine.  Simplified version with consistent node
+      // numbering with QuadInvIso is commented out below.
+      for ( int k{ 0 }; k < size1; ++k ) {
+        RealT phiA;
+        // NOTE: this limits this routine to quads
+        LinIsoQuadShapeFunc( xi1[0], xi1[1], k, phiA );
+        for ( int l{ 0 }; l < size1; ++l ) {
+          RealT phiB;
+          // NOTE: this limits this routine to quads
+          LinIsoQuadShapeFunc( xi1[0], xi1[1], l, phiB );
+          mortar_mat1[k * size1 + l] += phiA * phiB * quad_wt;
+        }
+      }
+
+      // 5. Evaluate mortar matrix (nonmortar/mortar contribs)
+      for ( int k{ 0 }; k < size1; ++k ) {
+        RealT phiA;
+        // NOTE: this limits this routine to quads
+        LinIsoQuadShapeFunc( xi1[0], xi1[1], k, phiA );
+        for ( int l{ 0 }; l < size2; ++l ) {
+          RealT phiB;
+          // NOTE: this limits this routine to quads
+          LinIsoQuadShapeFunc( xi2[0], xi2[1], l, phiB );
+          mortar_mat2[k * size2 + l] += phiA * phiB * quad_wt;
+        }
+      }
+
+      // Updated numbering version begin...
```

Comment:
Perhaps some additional clarificaiton on these commented lines is useful?

Reply:
Removed